[Metadata][TWDAPI-265] Added new Metadata API (#1175)
Samsung/tizen-docs: docs/application/web/guides/multimedia/metadata.md

> # Metadata

Reviewer: @nawab-samsung Please don't forget to guide to update toc_all.md file when a new page is added.

Reviewer 2: @pkosko 
1. Kindly correct the typo mistake of: 
#### [Ovierview](/application/web/guides/multimedia/overview.md)

Ovierview  --> Overview 

2. Kindly update the toc file.
https://github.com/Samsung/tizen-docs/blob/master/docs/application/toc_all.md

### Media and Camera
#### [Overview](/application/web/guides/multimedia/overview.md)
#### [JPEG File EXIF Information](/application/web/guides/multimedia/jpeg-exif.md)
#### [Audio Management](/application/web/guides/multimedia/audio.md)
#### [Audio Latency](/application/web/guides/multimedia/player-util.md)
#### [Media Controller](/application/web/guides/multimedia/media-controller.md)
#### [Media Key Events](/application/web/guides/multimedia/media-key.md)
#### [Radio](/application/web/guides/multimedia/radio.md)
#### [Camera](/application/web/guides/multimedia/camera.md)
#### [Metadata](/application/web/guides/multimedia/metadata.md)

You can extract metadata information from multimedia files.

The main features of the Metadata API include:

- Extraction of a metadata from a file

>   You can [extract metadata](#extract-metadata) from multimedia files using `get()` method.

Reviewer: From line 5 to here.
As there in only one item in the list, we can use a sentence form.

For example as like:

The main feature of the Metadata API is extraction of a metadata from a file. You can [extract metadata](#extract-metadata) from multimedia files using the `get()` method.

>   You can [extract metadata](#extract-metadata) from multimedia files using `get()` method.

Reviewer 2: @mijin85cho sure thanks.

## Prerequisites

> To access files using the Metadata API (in [mobile](../../api/latest/device_api/mobile/tizen/metadata.html), [wearable](../../api/latest/device_api/wearable/tizen/metadata.html)), the application has to request [proper storage permissions](../security/privacy-related-permissions.md).

Reviewer: When I check TWDAPI-265, it seems below privileges are requested.

http://tizen.org/privilege/mediastorage 
http://tizen.org/privilege/externalstorage 

If it is correct, would you please follow the style that used in other guide pages?

To access files using the Metadata API (in [mobile](../../api/latest/device_api/mobile/tizen/metadata.html),  [wearable](../../api/latest/device_api/wearable/tizen/metadata.html), and [TV]([wearable](../../api/latest/device_api/tv/tizen/metadata.html) ) applications), the application has to request permission by adding the following privileges to the `config.xml` file:
```
<tizen:privilege name="http://tizen.org/privilege/mediastorage "/>
<tizen:privilege name="http://tizen.org/privilege/externalstorage "/>
```

Or, is there any reason that you didn't follow the above form?

Author: Mentioned privileges are optional for applications. I added a section which describes this, besides of PPM-related part of comment.

> ## Extract metadata

Reviewer 2: The heading follows the latest prescribed sentence style.
@mijin85cho Should we change it to Title case to maintain consistency as conversion task will follow up to change the cases. 

## Extract Metadata

Reviewer: As this page is newly added one, let's use the new style that will be applied to whole pages.
I suggest to keep the current sentence style.

> 1. Create a file handle using `createFileHandle()` method:

Reviewer 2: ```suggestion
1. Create a file handle using the `createFileHandle()` method:

   ```javascript
   var filePath = "videos/sample_video.mp4";
   var fileHandle = tizen.metadata.createFileHandle(filePath);
   ```

> 2. When the file handle is properly created metadata can be extracted:

Reviewer 2: ```suggestion
2. When the file handle is properly created, metadata can be extracted:

   ```javascript
   console.log("Video duration is: " + fileHandle.get("DURATION") + " milliseconds");
   console.log("Video title is: " + fileHandle.get("TITLE"));
   ```

> 3. For better performance, FileHandle does not release underlying resources after every metadata extraction. On first extraction, data is decoded, thanks to that next metadata extractions are faster.

Reviewer 2: @mijin85cho request your opinion on the suggestion; if you do not like the suggestion, request you to provide your suggestion. 

I am suggesting changes because the sentence seems to be in colloquial language.

```suggestion
3. For better performance, the file handle does not release underlying resources after every metadata extraction. The data is decoded on the first extraction, which helps in speeding up the next metadata extractions.
```

@pkosko I have a doubt: 

Have you intentionally kept FileHandle case in line 31 as such, whereas the case differs  in line 17
 and 24.

In my suggestion, I have changed **FileHandle** to **file handle**. If the answer to above statement is **yes**, replace file handle with FileHandle then.

Reviewer: >I am suggesting changes because the sentence seems to be in colloquial language.

I think @nawab-samsung 's suggest to change the tone is good. Thanks for that.

> When all needed metadata are extracted, handle need to be released using `release()` method:

Reviewer 2: Kindly change are to is 
```suggestion
When all needed metadata is extracted, the file handle needs to be released using the`release()` method:

   ```javascript
   fileHandle.release();
   ```

## Related Information
- Dependencies
  - Tizen 6.0 and Higher for Mobile
  - Tizen 6.0 and Higher for Wearable

>   - Tizen 6.0 and Higher for TV

Reviewer 2: @pkosko 
I have a few doubts: 

- Is this feature also applicable to TV as well, as per line 11 its mobile & wearable only. 
- Is this feature supported from Tizen **6.0** and onwards?

Reviewer: >Is this feature also applicable to TV as well, as per line 11 its mobile & wearable only.

Good question.
@pkosko When I check https://code.sec.samsung.net/jira/browse/TWDAPI-265,
the TV is also marked to be supported.
So would you please check adding the API link to TV also on line 13?

>Is this feature supported from Tizen 6.0 and onwards?

I think that's why this PR is targeting tizen_6.0_prepare branch, not master branch.
If this update is merged into tizen_6.0_prepare, it will be kept on tizen_6.0_prepare for a while.
When Tizen 6.0 is released(Maybe on Oct.), every fixes on tizen_6.0_prepare will be merged into master branch and published.

Author: @nawab-samsung  and @mijin85cho
line: "To access files using the Metadata API (in [mobile](../../api/latest/device_api/mobile/tizen/metadata.html) and [wearable](../../api/latest/device_api/wearable/tizen/metadata.html) applications), the application has to request [proper storage permissions](../security/privacy-related-permissions.md)." 
says about necessary privilege request using PPM API. As TV does not have PPM feature support, I didn't mention it here, because using Metadata does not require storage privilege on TV.

# Metadata

You can extract metadata information from multimedia files.

The main feature of the Metadata API is to extract metadata from a file. You can [extract metadata](#extract-metadata) from multimedia files using the `get()` method.

> 

Reviewer: As the line 3 and line 5 are talking about the same content, how about removing line 3?

Also as we changed the second sentence into the overview sentence of this page,
it is better to give only simple description from overview while offering detailed description to the followed section.

So how about changing the sentence, for example, like below?

Metadata contains various information of multimedia files. You can [extract metadata](#extract-metadata) from multimedia files.

Author: @mijin85cho current functionality is a 'basic' part, which is planned to be extended in future. Used 'template' is consistent with other modules which uses one sentence of module summary, then the list of more detailed descriptions of features.

Planned features will  include: getting artwork, frames and synclyrics for video.
I agree that current description looks too simple, but for now the module is just as simple. We can use the form you proposed, which will be changed when the list of supported features will be extended, is it ok?

Reviewer 2: @mijin85cho 

Your suggestion is not contextually aligned (contains word slightly changes it)
How about this : 
Metadata is the information about multimedia files. You can extract metadata from multimedia files.

Reviewer: @pkosko I guess the planned features are also included in Tizen 6.0. 
As you said, I suggested sentence modification because of simple module. If you're planning to add additional features, for now let's keep the current.

Reviewer: > Your suggestion is not contextually aligned (contains word slightly changes it)
> How about this :
> Metadata is the information about multimedia files. You can extract metadata from multimedia files.

I found your comment after I replied to @pkosko. Sorry for the miss.

@pkosko One more suggestion. How about adding @nawab-samsung's suggested sentence to the overview?

Metadata is the information about multimedia files.

Reviewer 2: @pkosko can you please incorporate this comment, following which I will do a full review and approve if no more changes are required.

Author: It is done in line 3.

Reviewer 2: @pkosko  thanks a lot

> The main feature of the Metadata API is extraction of a metadata from a file. You can [extract metadata](#extract-metadata) from multimedia files using the `get()` method.

Reviewer 2: @pkosko Just to maintain writing style between pages, your sentence was also correct. 
```suggestion
The main feature of the Metadata API is to extract metadata from a file. You can [extract metadata](#extract-metadata) from multimedia files using the `get()` method.

## Prerequisites

To access files using the Metadata API (in [mobile](../../api/latest/device_api/mobile/tizen/metadata.html), [wearable](../../api/latest/device_api/wearable/tizen/metadata.html), and [tv](../../api/latest/device_api/tv/tizen/metadata.html) applications), the application has to request storage privileges by adding the following privileges to the `config.xml` file:

> 

Reviewer: > @mijin85cho I presume that the page is new one, two links that are not working will work afterwards

@nawab-samsung As I guided you before with https://github.com/Samsung/tizen-docs/pull/1076#discussion_r409391314,
you can change the "latest" in the address bar, to actual version (6.0 in this case) when you check on "View file" tab.

Please don't be confused that it is not asking you to guide the author to change the latest to 6.0,
as it is just a way for the reviewers to check the validation of the links to API.

Or, as pkosko guided, or guided on https://github.com/Samsung/tizen-docs/pull/1076#discussion_r409391314,
you can check the links on the staging URL.

Reviewer 2: @mijin85cho thanks, I do remember, but for me staging was down after multiple tries, when I tried to check so to ensure status of links, just asked @pkosko to verify the link but not change.

```xml
<!-- for accessing internal storage only -->
<tizen:privilege name="http://tizen.org/privilege/mediastorage"/>
<!-- for accessing external storage only -->
<tizen:privilege name="http://tizen.org/privilege/externalstorage"/>
```

> Additionally, to access files using the Metadata API (in [mobile](../../api/latest/device_api/mobile/tizen/metadata.html) and [wearable](../../api/latest/device_api/wearable/tizen/metadata.html) applications), the application has to request [proper storage permissions](../security/privacy-related-permissions.md) using the PPM API.

Reviewer: Suggestion. How about adding a link to PPM API?

... using the PrivacyPrivilege API (in \[mobile](../../api/latest/device_api/mobile/tizen/ppm.html) and \[wearable](../../api/latest/device_api/wearable/tizen/ppm.html) applications).

If I understood something wrong, please let me know.

Author: I agree, it is very good idea.

## Extract metadata

1. Create a file handle using the `createFileHandle()` method:

   ```javascript
   var filePath = "videos/sample_video.mp4";
   var fileHandle = tizen.metadata.createFileHandle(filePath);
   ```

2. When the file handle is properly created, metadata can be extracted:

   ```javascript
   console.log("Video duration is: " + fileHandle.get("DURATION") + " milliseconds");
   console.log("Video title is: " + fileHandle.get("TITLE"));
   ```

3. For better performance, the file handle does not release underlying resources after every metadata extraction. The data is decoded on the first extraction, which helps in speeding up the next metadata extractions.

> When all the needed metadata is extracted, the file handle needs to be released using the`release()` method:

Reviewer: ```suggestion
When all the needed metadata is extracted, the file handle needs to be released using the `release()` method:
```

You can extract metadata information from multimedia files.

The main feature of the Metadata API is extraction of a metadata from a file. You can [extract metadata](#extract-metadata) from multimedia files using the `get()` method.

## Prerequisites

> To access files using the Metadata API (in [mobile](../../api/latest/device_api/mobile/tizen/metadata.html), [wearable](../../api/latest/device_api/wearable/tizen/metadata.html) and [tv](../../api/latest/device_api/tv/tizen/metadata.html) applications), the application has to request storage privileges by adding the following privileges to the `config.xml` file:

Reviewer 2: KIndly add "," after **and** (Serial comma rule)
```suggestion 
To access files using the Metadata API (in [mobile](../../api/latest/device_api/mobile/tizen/metadata.html), [wearable](../../api/latest/device_api/wearable/tizen/metadata.html), and [tv](../../api/latest/device_api/tv/tizen/metadata.html) applications), the application has to request storage privileges by adding the following privileges to the `config.xml` file:

Reviewer 2: @pkosko  Kindly check the links. If Ok , ignore the comment.
@mijin85cho I presume that the page is new one, two links that are not working will work

Author: All links work well on staging page generated automatically:
https://docs1.stg.tizen.org/staging/1175/application/web/guides/multimedia/metadata/

Reviewer 2: @pkosko So sorry. you missed this

Kindly add "," after and (Serial comma rule)

https://github.com/Samsung/tizen-docs/pull/1175#discussion_r454187379

```xml
<!-- for accessing internal storage only -->
<tizen:privilege name="http://tizen.org/privilege/mediastorage"/>
<!-- for accessing external storage only -->
<tizen:privilege name="http://tizen.org/privilege/externalstorage"/>
```

> Additionally, to access files using the Metadata API (in [mobile](../../api/latest/device_api/mobile/tizen/metadata.html) and [wearable](../../api/latest/device_api/wearable/tizen/metadata.html) applications), the application has to request storage privileges [proper storage permissions](../security/privacy-related-permissions.md) using the PPM API.

Reviewer 2: @pkosko Do we need this ? this seems repetition. 

As per my understanding :
1. To access files using the Metadata API (in mobile, wearable and tv applications), the application has to request storage privileges by adding the following privileges to the config.xml file:
```
code
```

This seems fine. 

But the next part (line 19) seems to be redundant and only thing that differs is PPM API. Also, you have missed adding TV after wearable in line 19. 

Request you to kindly check, and if you still need line 19, then kindly add TV (If you have missed it).

I would suggest that we can remove it.

Author: This information is added on purpose. Storage privileges on mobile and wearable are privacy related, in this case:
1. Need to add privilege to config.xml
2. Use PPM API to `request` privilege from user in runtime (which results with popup asking user to access files). Word 'additionally' was used to show second action which needed.

2nd point is not needed on TV, so I omitted TV on purpose.

Reviewer 2: @pkosko We need slight changes to correct sentence structure. 

Kindly choose the appropriate one and verify for technical accuracy : 
```suggestion 
Additionally, to access files using the Metadata API (in [mobile](../../api/latest/device_api/mobile/tizen/metadata.html) and [wearable](../../api/latest/device_api/wearable/tizen/metadata.html) applications), the application has to request [storage privileges](../security/privacy-related-permissions.md) using the PPM API.

```
or 
```suggestion

Additionally, to access files using the Metadata API (in [mobile](../../api/latest/device_api/mobile/tizen/metadata.html) and [wearable](../../api/latest/device_api/wearable/tizen/metadata.html) applications), the application has to request [proper storage permissions](../security/privacy-related-permissions.md) using the PPM API.
```
